crates/core/src/operations/optimize.rs

```diff
@@ -606,44 +810,129 @@ impl MergePlan {
+                                async move {
+                                    let file_reader =
+                                        ParquetObjectReader::new(object_store_ref, meta.location)
+                                            .with_file_size(meta.size);
+                                    ParquetRecordBatchStreamBuilder::new(file_reader)
+                                        .await?
+                                        .build()
+                                }
+                            })
+                            .try_flatten()
+                            .boxed();
+
+                        let rewrite_result = tokio::task::spawn(Self::rewrite_files(
+                            task_params.clone(),
+                            partition,
+                            files,
+                            store.object_store(Some(operation_id)),
+                            futures::future::ready(Ok(batch_stream)),
+                        ));
+                        util::flatten_join_error(rewrite_result)
+                    } else {
+                        // === Compaction with global sort ===
+                        // Instead of simply merging files, we build a DataFusion plan
+                        // over the same set of files, apply a full global sort on the
+                        // user-specified `sort_columns`, and then rewrite pages in that
+                        // sorted order. This guarantees monotonic ordering across all
+                        // output files. Note: this is a standard DataFusion `sort(...)`
+                        // operation, not Z-ordering (see the ZOrder branch below).
+                        use crate::delta_datafusion::DataFusionMixins;
+                        use crate::delta_datafusion::DeltaScanConfigBuilder;
+                        use crate::delta_datafusion::DeltaTableProvider;
+                        use datafusion::execution::context::{SessionConfig, SessionContext};
+                        use datafusion::execution::memory_pool::FairSpillPool;
+                        use datafusion::execution::runtime_env::RuntimeEnvBuilder;
+                        use datafusion::logical_expr::col;
+                        use url::Url;
+
+                        // Use cloned store for DataFusion operations
+                        let log_store_ref = store.clone();
+                        let snapshot = snapshot.clone();
+                        let sort_cols = self.sort_columns.clone();
+                        // Extract underlying file list from MergeBin
+                        let files_for_provider = files.files.clone();
+                        // Asynchronous DataFrame-based sorted reader
+                        let read_sorted = async move {
+                            let memory_pool = FairSpillPool::new(max_spill_size);
+                            let runtime = RuntimeEnvBuilder::new()
```

Comment:
You shouldn't need this machinery shouldn't session ctx and proper object stores already be registered?